Docs: Technical Design (#73)
Greater-Wellington-Regional-Council/Environmental-Outcomes-Platform: docs-site/src/markdown-pages/design/technical-design.md

  and other systems that provide services that EOP can use.

## Structure

EOP is being developed as a hub and spoke model. Data is captured or acquired
from external sources and sent to the hub in a raw format. That raw format can
then be forwarded to different storage or processing components depending on how
that data will be used, for example, it might be stored in a data warehouse for
analytical analysis or to a real time processing component for operational
alerting. Key to this model is not forcing all data into a store which would
limit its to be used for different use cases.

A decision has been made to build the hub on [Kafka](https://kafka.apache.org/).
Kafka was chosen to provide a scalable, loosely coupled, hub with well supported
connectors to other systems. While Kafka is often used for real time processing
it is also well suited to consuming data that may come from batch processes from
the councils.

The following diagram shows the high-level relationship between the different
classes of components that will be delivered around the hub

![High Level Overview Diagram](./high-level-overview.png)

### Acquisition

> Acquisition components support the process of getting data captured in existing
> systems and into EOP. These might be existing council systems or any third party
> systems which EOP can make use of. Acquiring data in this fashion will be the
> main focus of the initial EOP build.
> 
> Generally, the preference is for councils to push data into EOP rather than
> pull. This allows councils to choose the frequency of updates from their systems
> to EOP. However, there will be some cases where pulling data from a system is
> required, for example from third party systems that can't be changed.
> 
> A special case is Hilltop which because of its ubiquity in councils and it
> having a pre-existing API will be supported as a "pull from councils" example.
> 
> Example Acquisition components:
> 
> - JSON API's exposed to councils
> - Data lake / Blob storage style end points for bulk data
> - Hilltop API
> - Direct connection from Kafka Connectors running in council environments

Reviewer: ```suggestion
Acquisition components support ingestion of data from existing systems into EOP.
These might be existing council systems or third party systems housing data
which EOP can make use of. Acquisition will be the the initial focus for EOP.

The preference is for councils to push data to EOP, rather than for EOP to pull.
This allows councils to control the frequency of updates. However, there will be
some cases where pulling data from a system is required, such as data from third
party systems that can't be changed.

A special case is Hilltop. Because of its wide-use and existing APIs, EOP will
support pulling data from council Hilltop instances.

Example Acquisition components:

- JSON API's exposed to councils
- Data lake / Blob storage style end points for bulk data
- Hilltop API
- Direct connection from Kafka Connectors running in council environments
```

### Capture

> Capture components allow for getting information where EOP is the system of
> record. This will be used when there is a potential for providing new tools to
> all councils for replacing existing data capture methods. New tooling will allow
> for bring the data captured up to new standards for meta-data that are not
> caught in existing systems.
> 
> This type of component is a long term vision for the EOP solution.
> 
> Example Capture components:
> 
> - Field capture user interface that may be running on a mobile device
> - IOT endpoints for capturing data from sensors
> - User interface for QA review

Reviewer: ```suggestion
Capture components support ingestion of data where EOP is the system of record.
These will be developed when there is an opportunity to improve on existing data
capture tools for all councils, such as supporting newer data-standards or
capturing meta-data not-supported by existing systems.

This type of component is a longer term goal for the EOP.

Example Capture components:

- Field capture user interface that may be running on a mobile device
- IOT endpoints for capturing data from sensors
- User interfaces for QA review
```

## Data Stores

> Data store components are about storing data in a way that is appropriate for
> the type of analysis being performed. The data storage type being used drives
> the types of queries that can be performed and the type of information that can
> be derived. By using the hub and spoke model with Kafka as the hub, EOP can stay
> agnostic of the data stores that are used, rahter than tying the solution to a
> specific type of analysis.
> 
> Example Data store components:
> 
> - OLTP database for fast operational access
> - OLAP database for slower analytical access
> - Graph database for network analysis

Reviewer: ```suggestion
Multiple data store components may be used to allow modelling and querying based
on the type analysis required.

Using a Hub and Spoke model allows the EOP to remain agnostic of the data store 
solutions used now and in the future.

Example Data store components:

- OLTP database for fast operational access
- OLAP database for slower analytical access
- Graph database for network analysis
```

## Processing

> Processing components are there to perform analysis on the data that is stored
> in EOP.
> 
> The processing components are loosely coupled to the data stores. The intention
> is for this to be a key extension point for adding new analysis processes to EOP
> without changing other components.
> 
> Example Processing components:
> 
> - Real time processing has driven off kafka streams
> - Batch processing driven data in the OLAP data stores
> - Data science scripts (R / Python) for analysis e.g. Naturalised flow
> - AI model training

Reviewer: ```suggestion
Processing components perform analysis on the data stored in EOP.

These are decoupled from the data stores, and designed to be a key extension
point for adding new analysis capability without affecting other components.

Example Processing components:

- Real time processing, driven from kafka streams
- Batch processing, driven from data in OLAP data stores
- Data science scripts (R / Python) for analysis e.g. Naturalised flow
- AI model training
```

## Output

> Output components systems take information captured in EOP and present it to
> users as usable outcomes. The diagram shows the output components grouped by the
> type of user that they are intended for. That is to draw attention to the fact
> that delivering an outcome may be presented differently depending on the
> consumers even though they may be in the same type of technology.
> 
> Example Processing components:
> 
> - UI Viewers
> - API endpoints
> - Bulk data downloads
> 
> ### External Applications B2C
> 
> Applications that general public are intended to consume, for these there needs
> to be a focus on the user experience making the tools valuable for the user.
> 
> ### External Applications B2B
> 
> Applications that are intended to be used by third party organisations including
> central government. The information returned may have a focus on providing for
> specific questions. These types of application may be more focused on
> information exported for use in third party systems
> 
> ### Internal Tools
> 
> Tools that are used by council staff to perform their day to day work. These
> applications need a user experience that is task focused for the type of person
> using them and may expect a higher level of expertise when understanding the
> data.
> 
> ### Feeds to Council Systems
> 
> Feeds to council systems where EOP has created information that is then fed back
> to existing council systems for presentation

Reviewer: ```suggestion
Output components present information captured and generated by EOP to various
types of users, with a focus on outcomes.

The diagram shows these grouped by the type of user emphasise a user-centric,
rather than technology-centric approach. E.g. multiple web applications or APIs
may be developed to serve difference audiences.

 Example Processing components:

- UI Viewers
- API endpoints
- Bulk data downloads

### External Applications B2C

Applications intended for public use. These should be focussed on presenting 
information and outcomes in a user-friendly way.

### External Applications B2B

Applications intended for use by third-party organisations including central
government. This could be information for particular types of reporting, often
provided as exports for use in third party systems.

### Internal Tools

Tools used by council staff to perform day-to-day work. These might task-focussed
applications, developed for users with particular expertise in the data.

### Feeds to Council Systems

These may be useful where EOP has created information that is valuable for
presentation by existing council systems.
```
